pyrobosim/pyrobosim/navigation/astar_grid.py

```diff
@@ -0,0 +1,329 @@
+                waypoints.append(curr_start)
+                curr_dx, curr_dy = new_dx, new_dy
+                curr_start = poses[i - 1]
+        return waypoints
+
+    def _generate_path(self, node):
+        """
+        Backtrack from a node to generate the full path till that node
+        :param node: The node from which to retrace the path
+        :type node: :class: Node
+        """
+        poses = []
+        while node is not None:
+            x, y = self.grid.grid_to_world(node.x, node.y)
+            poses.append(Pose(x=x, y=y))
+            node = node.parent
+        poses.reverse()
+        print(f"Number of waypoints in full path : {len(poses)}")
+        poses = self._reduce_waypoints(poses)
+        print(f"Number of waypoints in reduced path : {len(poses)}")
+        self.latest_path = Path(poses=poses)
+        self.latest_path.fill_yaws()
+
+    def _is_valid_start_goal(self):
+        """Validate the start and goal locations provided to the planner"""
+        valid = True
+        if self.grid.is_occupied(self.start[0], self.start[1]):
+            warnings.warn(f"Start position {self.start} is occupied")
+            valid = False
+        if self.grid.is_occupied(self.goal[0], self.goal[1]):
+            valid = False
+            warnings.warn(f"Goal position {self.goal} is occupied")
+        return valid
+
+    def plan(self, start, goal):
+        """
+        Plans a path from start to goal
+        :param start: The start postion in world coordinates
+        :type start: :class: Pose
+        :param goal: The goal postion in world coordinates
+        :type goal: :class: Pose
+        """
+        self._reset()
+        self.start = self.grid.world_to_grid(start.x, start.y)
+        self.goal = self.grid.world_to_grid(goal.x, goal.y)
+
+        # If start or goal position is occupied, return empty path with warning.
+        if not self._is_valid_start_goal():
+            return Path()
+
+        start_node = Node(
+            self.start[0], self.start[1], g=0, h=self._heuristic(self.start, self.goal)
+        )
+        goal_node = Node(self.goal[0], self.goal[1], 0, 0)
+
+        print("Planning : Started")
+        timed_out = False
+        path_found = False
+        t_start = time.time()
+        self.candidates.put(start_node)
+        while not path_found and self.candidates and not timed_out:
+            current = self._get_best_candidate()
+            if current == goal_node:
+                path_found = True
+            else:
+                self._expand(current)
+            self._mark_visited(current)
+            self.planning_time = time.time() - t_start
+            timed_out = self.planning_time >= self.max_time
+
+        if path_found:
+            print("Planning : Success")
+            self._generate_path(current)
+            return self.latest_path
+        else:
+            print("Planning : Failed")
+            return Path()
+
+    def print_metrics(self):
```

Comment:
For some of the other things you had debug prints for, like the heuristic used and allowing diagonal moves, you may feel free to move them to this function as you'd like. I'm fine either way